src/include/function/gds/gds_frontier.h

```diff
@@ -0,0 +1,140 @@
+ * to determine if the algorithm has converged or not.
+ *
+ * All functions supported in this base interface are thread-safe.
+ */
+class Frontiers {
+
+public:
+    explicit Frontiers(uint64_t initialActiveNodes) {
+        numNextActiveNodes.store(initialActiveNodes);
+        curIter.store(UINT64_MAX);
+    }
+    // TODO(Reviewer): Should I do this? I'm doing it because the compiler gives me this warning
+    // "warning: destructor called on non-final 'kuzu::function::PathLengthsFrontiers' that has
+    // virtual functions but non-virtual destructor [-Wdelete-non-abstract-non-virtual-dtor]"
+    // According to a Google search, I should add a virtual destructor but I don't fully understand
+    // if the deriving class, so PathLengthsFrontiers in this case, also needs a destructor.
+    virtual ~Frontiers() {};
+    virtual bool getNextFrontierMorsel(RangeFrontierMorsel& frontierMorsel) = 0;
+    void incrementNextActiveNodes(uint64_t i) { numNextActiveNodes.fetch_add(i); }
+    void beginNewIterationOfUpdates() {
+        lock_t lck{mtx};
+        // If curIter is UINT64_MAX, which indicates that the iterations have not started, the
+        // following line will set it to 0.
+        curIter.fetch_add(1);
+        numNextActiveNodes.store(0u);
+        beginNewIterationOfUpdatesInternalNoLock();
+    }
+    // When performing computations on multi-label graphs, it may be beneficial to fix a single
+    // node table of nodes in the current frontier and a single node table of nodes for the next
+    // frontier. That is because algorithms will generally perform extensions using a single
+    // relationship table at a time, and each relationship table R is between a single source node
+    // table S and a single destination node table T. Therefore, generally during execution the
+    // algorithm will need to check only the active S nodes in current frontier and update the
+    // active statuses of only the T nodes in the next frontier. The information that the algorithm
+    // is beginning and S-to-T extensions can be given to the frontiers to possibly avoid them
+    // doing lookups of S and T-related data structures, e.g., masks, internally.
+    // Note that ideally a similar function, say called "fixNodeTable", could be in GDSFrontier,
+    // since Frontiers has two GDSFrontier instances in it. However, some implementations of
+    // Frontiers can use a single GDSFrontier to mimic two frontiers (e.g., identifying nodes
+    // in current or next frontier by specific mask values. For such implementations we would have
+    // to call fixNodeTable(S) and fixNodeTable(T) on the same object, which could overwrite
+    // each other. That is why this function is put in the Frontiers class.
+    virtual void beginFrontierUpdatesBetweenTables(table_id_t curFrontierTableID,
+        table_id_t nextFrontierTableID) = 0;
+    virtual GDSFrontier* getCurFrontier() = 0;
```

Comment:
What about saving two GDSFrontier in this class as current and next so that we don't need to make these getters pure virtual.

Reply:
Done.